src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4857,10 +4857,22 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
 
         private static string GetChildNameFromPsObject(dynamic psObject, char separator)
         {
-            // The obvious solution would be to use the "PSChildName" property
-            // but some providers don't have it (like the Variable provider)
-            // So we use a substring of "PSPath" instead.
-            string childName = psObject.PSPath ?? string.Empty;
+            if (((PSObject)psObject).BaseObject is string result)
```

Comment:
Have you considered `PSObject.Base(psObject)`?

Reply:
No I didn't consider that. I've looked at the `Base` method and it does some unnecessary checks that makes it slower than just casting it. And I think the cast is needed because how else would I check that it's a string?

Reply:
> how else would I check that it's a string?

psObject is dynamic.

Reply:
Elaborate please, I don't understand how the code would be written. Are you saying that: `psObject is string result` (where psObject is the variable name) would work the same way as the current code? My understanding is that I need to unwrap the PSObject before I can do the type check.

Reply:
I think `if ((psObject.BaseObject is string result)` should work because old code was `string childName = psObject.PSPath ?? string.Empty;` - without cast.

Reply:
I just tested it and it does not seem to work. I guess it's because the dynamic implementation on PSObject only looks for members on the unwrapped object with no fallback to the original PSObject properties. My `psObject is string result` idea also doesn't work.

Reply:
> I just tested it and it does not seem to work. 

I wonder that `string childName = psObject.PSPath ?? string.Empty;` work. Exactly works? Or it is bug?
Maybe follow works:
```c#
object tmp = psObject.BaseObject;
if (tmp is string result)
...
```

I don't understand difference between `psObject.PSPath` and `psObject.BaseObject`. `PSPath` says that `psObject` is _always_ PSObject. Why do we define it as dynamic?
I tend to think that this code either doesn't work correctly or it's too gimmicky and at the very least needs refactoring.

Reply:
> I wonder that string childName = psObject.PSPath ?? string.Empty; work. Exactly works? Or it is bug?
The null check can probably be removed. I don't remember why I put it there but if it's null then there's something wrong so it's better to just throw.

> I don't understand difference between psObject.PSPath and psObject.BaseObject. PSPath says that psObject is always PSObject. Why do we define it as dynamic?
It's so we can access the dynamic members more easily. We could make it a PSObject and use the appropriate properties/methods to access those members but the old code used dynamic so I figured it was fine to use here during the rewrite.